xbmc/windowing/osx/WinEventsOSXImpl.mm

```diff
@@ -0,0 +1,365 @@
+      newEvent.button.button = XBMC_BUTTON_RIGHT;
+      newEvent.button.x = location.x;
+      newEvent.button.y = location.y;
+      [self MessagePush:&newEvent];
+      break;
+    case kCGEventRightMouseDown:
+      newEvent.type = XBMC_MOUSEBUTTONDOWN;
+      newEvent.button.button = XBMC_BUTTON_RIGHT;
+      newEvent.button.x = location.x;
+      newEvent.button.y = location.y;
+      [self MessagePush:&newEvent];
+      break;
+    case kCGEventOtherMouseUp:
+      newEvent.type = XBMC_MOUSEBUTTONUP;
+      newEvent.button.button = XBMC_BUTTON_MIDDLE;
+      newEvent.button.x = location.x;
+      newEvent.button.y = location.y;
+      [self MessagePush:&newEvent];
+      break;
+    case kCGEventOtherMouseDown:
+      newEvent.type = XBMC_MOUSEBUTTONDOWN;
+      newEvent.button.button = XBMC_BUTTON_MIDDLE;
+      newEvent.button.x = location.x;
+      newEvent.button.y = location.y;
+      [self MessagePush:&newEvent];
+      break;
+    case kCGEventMouseMoved:
+    case kCGEventLeftMouseDragged:
+    case kCGEventRightMouseDragged:
+    case kCGEventOtherMouseDragged:
+      newEvent.type = XBMC_MOUSEMOTION;
+      newEvent.motion.x = location.x;
+      newEvent.motion.y = location.y;
+      [self MessagePush:&newEvent];
+      break;
+    case kCGEventScrollWheel:
+      // very strange, real scrolls have non-zero deltaY followed by same number of events
+      // with a zero deltaY. This reverses our scroll which is WTF? anoying. Trap them out here.
+      if (nsevent.deltaY != 0.0)
+      {
+        newEvent.type = XBMC_MOUSEBUTTONDOWN;
+        newEvent.button.x = location.x;
+        newEvent.button.y = location.y;
+        newEvent.button.button =
+            CGEventGetIntegerValueField(event, kCGScrollWheelEventDeltaAxis1) > 0
+                ? XBMC_BUTTON_WHEELUP
+                : XBMC_BUTTON_WHEELDOWN;
+        [self MessagePush:&newEvent];
+
+        newEvent.type = XBMC_MOUSEBUTTONUP;
+        [self MessagePush:&newEvent];
+      }
+      break;
+
+    // handle keyboard events and transform them into the xbmc event world
+    case kCGEventKeyUp:
+      newEvent = [self keyPressEvent:&event];
+      newEvent.type = XBMC_KEYUP;
+
+      [self MessagePush:&newEvent];
+      passEvent = false;
+      break;
+    case kCGEventKeyDown:
+      newEvent = [self keyPressEvent:&event];
+      newEvent.type = XBMC_KEYDOWN;
+
+      if (![self ProcessOSXShortcuts:newEvent])
+        [self MessagePush:&newEvent];
+      passEvent = false;
+
+      break;
+    default:
+      return nsevent;
+  }
+  // We must return the event for it to be useful if not already handled
+  if (passEvent)
+    return nsevent;
+  else
+    return nullptr;
+}
+
+- (XBMC_Event)keyPressEvent:(CGEventRef*)event
+{
+  UniCharCount actualStringLength = 0;
+  // Get stringlength of event
+  CGEventKeyboardGetUnicodeString(*event, 0, &actualStringLength, NULL);
+
+  // Create array with size of event string
+  UniChar unicodeString[actualStringLength];
```

Comment:
What about using std::vector instead?

Reply:
Yeah, its just not worth it. UniChar type is expected for when its used. Casting in and out just isnt worth it imo

Reply:
you don't need to cast vector, simply pass `vec.data()`